internal/core/command/controller/messaging/internal.go

```diff
@@ -0,0 +1,74 @@
+//
+// Copyright (C) 2022 IOTech Ltd
+//
+// SPDX-License-Identifier: Apache-2.0
+
+package messaging
+
+import (
+	"context"
+	"strings"
+
+	bootstrapContainer "github.com/edgexfoundry/go-mod-bootstrap/v2/bootstrap/container"
+	"github.com/edgexfoundry/go-mod-bootstrap/v2/di"
+	"github.com/edgexfoundry/go-mod-core-contracts/v2/errors"
+	"github.com/edgexfoundry/go-mod-messaging/v2/pkg/types"
+
+	"github.com/edgexfoundry/edgex-go/internal/core/command/container"
+)
+
+func SubscribeCommandResponses(ctx context.Context, dic *di.Container) errors.EdgeX {
+	lc := bootstrapContainer.LoggingClientFrom(dic.Get)
+	messageBusInfo := container.ConfigurationFrom(dic.Get).MessageQueue
+	internalResponseTopic := messageBusInfo.Internal.Topics[ResponseTopic]
+	externalResponseTopicPrefix := messageBusInfo.External.Topics[ResponseCommandTopicPrefix]
+
+	messages := make(chan types.MessageEnvelope)
+	messageErrors := make(chan error)
+	topics := []types.TopicChannel{
+		{
+			Topic:    internalResponseTopic,
+			Messages: messages,
+		},
+	}
+
+	messageBus := bootstrapContainer.MessagingClientFrom(dic.Get)
+	err := messageBus.Subscribe(topics, messageErrors)
+	if err != nil {
+		return errors.NewCommonEdgeXWrapper(err)
+	}
+
+	qos := messageBusInfo.External.QoS
+	retain := messageBusInfo.External.Retain
+	externalMQTT := bootstrapContainer.ExternalMQTTMessagingClientFrom(dic.Get)
+	go func() {
+		for {
+			select {
+			case <-ctx.Done():
+				lc.Infof("Exiting waiting for MessageBus '%s' topic messages", internalResponseTopic)
+				return
+			case err := <-messageErrors:
+				lc.Error(err.Error())
+			case msgEnvelope := <-messages:
+				lc.Debugf("Command response received on message queue. Topic: %s, Correlation-id: %s ", internalResponseTopic, msgEnvelope.CorrelationID)
+
+				// expected internal command response topic scheme: #/<service-name>/<device-name>/<command-name>/<method>
+				topicLevels := strings.Split(msgEnvelope.ReceivedTopic, "/")
+				length := len(topicLevels)
+				if length < 4 {
+					lc.Error("Failed to parse and construct command response topic scheme, expected request topic scheme: '#/<service-name>/<device-name>/<command-name>/<method>'")
+					continue
+				}
+
+				// expected external command response topic scheme: #/<device-name>/<command-name>/<method>
+				deviceName := topicLevels[length-3]
+				commandName := topicLevels[length-2]
+				method := topicLevels[length-1]
+				externalResponseTopic := strings.Join([]string{externalResponseTopicPrefix, deviceName, commandName, method}, "/")
+				publishMessage(externalMQTT, externalResponseTopic, qos, retain, msgEnvelope, lc)
```

Comment:
@lenny-intel do you think we should check the request id and publish it only if core-command really received the request?

Reply:
Yes, I had been thinking about this. We have to save active requests in two maps using requestId as keys. One map for external requests and one map for internal request (i.e. from an App Service). This allows code to know where to route the response (to external MQTT or back to internal MessageBus)

Figured this would be an handled once the code is handling internal request, but should start with the map for external request now.